### PR TITLE
refactor toobj.c mangle()

### PR DESCRIPTION
**Pull request by WalterBright**
1. pushes undocumented side effect of mangle() into caller so function becomes pure and eliminates a parameter
2. improve comments
3. eliminate memory leak by using peekString instead of extractString
4. reorganize some lines so code appears as coherent chunks instead of disorganized tangle
5. use Ddoc style function comments

### CODE REVIEW

File: src/toobj.c [hunk 1]
```diff
@@ -1288,43 +1288,48 @@ void toObjFile(Dsymbol *ds, bool multiobj)
...
 
-            const char *tlvInitName = buffer.extractString();
-            Symbol *tlvInit = symbol_name(tlvInitName, SCstatic, type_fake(vd->type->ty));
+            Symbol *tlvInit = symbol_name(tlvInitName, SCstatic, t);
             tlvInit->Sdt = NULL;
             tlvInit->Salignment = type_alignsize(s->Stype);
-
-            type_setty(&tlvInit->Stype, tlvInit->Stype->Tty | mTYthreadData);
-            type_setmangle(&tlvInit->Stype, mangle(vd, tlvInit));
+            if (vd->linkage == LINKcpp)
```
Comment (ZombineDev) [inline src/toobj.c]: Does LINKcpp include extern (C)? If not, you may need to also check for C linkage a C11 also has TLS support (http://en.cppreference.com/w/c/thread/thread_local).

Comment (WalterBright) [inline src/toobj.c]: I didn't write that code, I don't know why this is special cased for LINKcpp. Changing that behavior should be in a separate PR anyway.
(LINKc is distinct from LINKcpp)

Comment (ZombineDev) [inline src/toobj.c]: Alright, thanks for the clarification.

Comment (jacob-carlborg) [inline src/toobj.c]: I'm pretty sure I copied the code for setting the mangling from somewhere else. Here are two cases [1][2] which sets the mangling and setting `Sflags` to `|= SFLpublic` for the C++ case.

[1] https://github.com/dlang/dmd/blob/master/src/tocsym.c#L250
[2] https://github.com/dlang/dmd/blob/master/src/tocsym.c#L347

Comment (WalterBright) [inline src/toobj.c]: Of course, we could disable that and see what breaks, but that's beyond the scope of this PR.

Comment (jacob-carlborg) [inline src/toobj.c]: Absolutely, just trying to explain why the code looks like it does.

**Comment by andralex** [comment]
Auto-merge toggled on
